### PR TITLE
[Sol->Yul] Member access for structs.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 1]
```diff
@@ -1467,7 +1467,43 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
 		break;
 	case Type::Category::Struct:
 	{
-		solUnimplementedAssert(false, "");
+		StructType const& structType = dynamic_cast<StructType const&>(*_memberAccess.expression().annotation().type);
```
Comment (mijovic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
		auto const& structType = dynamic_cast<StructType const&>(*_memberAccess.expression().annotation().type);
```

**Review by mijovic — COMMENTED** [review]
Are there more tests to be activated?

**Comment by mijovic** [comment]
> Are there more tests to be activated?

I added a commit that activates additional tests.

**Comment by mijovic** [comment]
> > Are there more tests to be activated?
> 
> I added a commit that activates additional tests.

Even after adding more tests `semanticTests/various/negative_stack_height.sol` fails with `StackTooDeepError`.

**Comment by chriseth** [comment]
Can you check if we get `StackTooDeep` only because the optimizer is not activated?

**Comment by mijovic** [comment]
> Can you check if we get `StackTooDeep` only because the optimizer is not activated?

It is trying to compile it with viaYul, and it comes from `optimize()` call in `AssemblyStack`. Still, that error should be caught in SemanticTest and it should be fine as test is marked with compileViaYul: false ...

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 2]
```diff
@@ -1467,7 +1467,43 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
...
+
+		IRVariable expression(_memberAccess.expression());
+		switch (structType.location())
+		{
+		case DataLocation::Storage:
+		{
+			pair<u256, unsigned> const& offsets = structType.storageOffsetsOfMember(member);
```
Comment (ekpyron) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1477]:
```suggestion
			auto const& [slotOffset, offsetInSlot] = structType.storageOffsetsOfMember(member);
```

**Review by ekpyron — CHANGES_REQUESTED** [review]
One minor suggestion plus @mijovic's earlier suggestion - otherwise this looks good!

**Comment by mijovic** [comment]
There are more tests to be active. I remember I added commit with those, but in rebase/force push it was lost probably :)

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 2 above)
Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1477]: This one is still open...

**Review by ekpyron — APPROVED** [review]

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 2 above)
Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1477]: But well, I'd call it "optional" and am approving.

**Review by mijovic — APPROVED** [review]

**Pull request merged**